soc/usbexecd.c

```diff
@@ -0,0 +1,598 @@
+// Param: id - which process is sending this
+// Param: arg - an argument for the command (like ACK)
+// Param len - the length of the data being sent following this packet
+void send_header(uint8_t cmd, uint8_t id, uint8_t arg, uint8_t len) {
+    // Create the packet
+    uint8_t buf[4] = {cmd, id, arg, len};
+    // Write the packet to the socket
+    int r = write(sock, buf, sizeof(buf));
+    // Ensure it wrote all of the bytes
+    if (r != sizeof(buf)) {
+        fatal("send_header failed: %s", strerror(errno));
+    }
+}
+
+void read_until(int fd, void *buf, int len) {
+    int total_read = 0;
+    int single_read = 0;
+
+    while (total_read < len) {
+        single_read = read(fd, buf, len-total_read);
+
+        if (single_read < 0) {
+            fatal("Unable to read from pipe in synchronous read loop");
+        }
+
+        total_read+=single_read;
+    }
+}
+
+// Constructor function for both read-only and write-only pipe buffers
+int pipebuf_common_init(pipebuf_t* pb, int id, int role, uint32_t events, unsigned bufend) {
+    // Assign all fields to the pipe buffer
+    pb->id = id;
+    pb->role = role;
+    pb->eof = false;
+    pb->startpos = pb->endpos = pb->bufcount = 0;
+    pb->credit = 0;
+
+    // Allocate memory for file descriptors
+    int pipefd[2];
+    // Create the pipes
+    int r = pipe(pipefd);
+    // Verify it worked
+    if (r < 0) {
+        fatal("pipe failed: %s", strerror(errno));
+    }
+
+    // Set the proper read or write pipe file descriptor
+    pb->fd = pipefd[bufend];
+
+    // Create an epoll_event related to that file descriptor
+    struct epoll_event evt;
+    evt.data.ptr = pb;
+    evt.events = events;
+
+    // Register the event
+    r = epoll_ctl(epfd, EPOLL_CTL_ADD, pb->fd, &evt);
+    // Verify it worked
+    if (r < 0) {
+        fatal("epoll_ctl failed: %s", strerror(errno));
+    }
+
+    // Return the other file descriptor?
+    return pipefd[!bufend];
+}
+
+// Closing functionality for read and write pipe buffers
+void pipebuf_common_close(pipebuf_t* pb) {
+    // If the file descriptor exists (which it should)
+    if (pb->fd) {
+        // Close it
+        close(pb->fd);
+        // Set it back to 0
+        pb->fd = 0;
+    }
+    // Reset other fields
+    pb->startpos = pb->endpos = pb->bufcount = 0;
+    pb->eof = true;
+
+    // Tell the other end of the pipe that this fd has been closed
+    send_header(CMD_CLOSE_CONTROL + pb->role, pb->id, 0, 0);
+}
+
+int pipebuf_in_init(pipebuf_t* pb, int id, int role) {
+    return pipebuf_common_init(pb, id, role, EPOLLIN, 0);
+}
+
+// We have data coming in to stdout or stderr
+// and need to write it to the CLI
+void pipebuf_in_handle(pipebuf_t* pb) {
+    int len = 255;
+    // if (pb->credit < len) len = pb->credit;
+    char buf[255];
+    len = read(pb->fd, buf, len);
+    if (len < 0) {
+        if (errno == EAGAIN) {
+            return;
+        }
+        fatal("error in pipe read: %s", strerror(errno));
+    }
+
```

Comment:
If `len == 0`, you've detected EOF and should signal the close event to the CLI and close the socket.